flang/lib/Lower/IO.cpp

```diff
@@ -921,6 +921,105 @@ constexpr bool isDataTransferNamelist<Fortran::parser::PrintStmt>(
+                           fir::CharacterType::get(builder.getContext(), 1)));
+  auto symbol = GetLastSymbol(&expr);
+  auto labelIter = assignMap.find(*symbol);
+  assert(labelIter != assignMap.end() && "ICE: Label not found in map");
+  auto labels = labelIter->second;
+
+  for (auto label : labels) {
+    indexList.push_back(label);
+    auto iter = labelMap->find(label);
+    assert(iter != labelMap->end() && "ICE: Label is missing from the table");
+
+    // Extract the inner text from the format statement
+    auto text = toStringRef(iter->second->position);
+    text = text.drop_front(text.find('('));
+    text = text.take_front(text.rfind(')') + 1);
+
+    // Lower the format string as a string literal and get the reference to it
+    auto formatString = fir::getBase
+      (createStringLiteral(loc, converter, text, text.size()));
+    auto stringLen = builder.create<mlir::ConstantIntOp>
+      (loc, text.size(), builder.getI64Type());
+    auto formatStringConv = builder.createConvert
+      (loc, stringRefType, formatString);
+
+    // Pass the format string reference and the string length out of the select
+    // statement
+    llvm::SmallVector<mlir::Value, 8> args;
+    args.push_back(formatStringConv);
+    args.push_back(stringLen);
+    builder.create<mlir::BranchOp>(loc, endBlock, args);
+
+    // Add block to the list of cases and make a new one
+    blockList.push_back(block);
+    block = block->splitBlock(builder.getInsertionPoint());
+    builder.setInsertionPointToEnd(block);
+  }
+
+  // Create the unit case which should result in an error
+  auto* unitBlock = block->splitBlock(builder.getInsertionPoint());
+  builder.setInsertionPointToEnd(unitBlock);
+
+  // TODO: Replace with instructions to crash the program
+  auto emptyString = fir::getBase(createStringLiteral(loc, converter, "", 0));
+  auto emptyStringConv = builder.createConvert(loc, stringRefType, emptyString);
+  auto zero = builder.create<mlir::ConstantIntOp>(loc, 0, builder.getI64Type());
+  llvm::SmallVector<mlir::Value, 8> args;
+  args.push_back(emptyStringConv);
+  args.push_back(zero);
+  builder.create<mlir::BranchOp>(loc, endBlock, args);
+
+  // Add unit case to the select statement
+  blockList.push_back(unitBlock);
+
+  // Lower the selectOp
+  builder.setInsertionPointToEnd(startBlock);
+  auto label = Fortran::lower::createSomeExpression(loc, converter,
+                                                    expr, localSymbols);
+  builder.create<fir::SelectOp>(loc, label, indexList, blockList);
+
+  builder.setInsertionPointToEnd(endBlock);
+  endBlock->addArgument(stringRefType);
+  endBlock->addArgument(builder.getI64Type());
+
+  // Handle and return the string reference and length selected by the selectOp
+  auto stringRef = endBlock->getArgument(0);
+  auto stringLen = endBlock->getArgument(1);
+  auto helper = Fortran::lower::CharacterExprHelper{builder,loc};
+  auto charBox = helper.toExtendedValue(stringRef, stringLen).getCharBox();
```

Comment:
You probably did this for a reason I do not see, but I do not get why you need to create the extended value here instead of returning `{stringRef, stringLen, mlir::Value{}}`.

Reply:
You are right, I forgot to refactor it out.